feat: add request cancellation functionality

sasanktumpati opened the pull request:
## PR Description

Added request cancellation functionality.

## Related Issues

- #109

### Checklist
- [x] I have gone through the [contributing guide](https://github.com/foss42/apidash/blob/main/CONTRIBUTING.md)
- [x] I have updated my branch and synced it with project `main` branch before making this PR
- [x] I am using the latest Flutter stable branch (run `flutter upgrade` and verify)
- [x] I have run the tests (`flutter test`) and all tests are passing

## Added/updated tests?

- [ ] Yes
- [x] No, and this is why: Implementation focuses on request cancellation functionality.

## OS on which you have developed and tested the feature?

- [ ] Windows
- [ ] macOS
- [x] Linux

## Changes

### HTTP Client Management
- Added handling for `http.ClientException` in request cancellation scenarios
- Implemented client instance management per request ID

### UI Updates
- Modified SendButton to toggle between Send/Cancel states
- Updated ResponsePane to show cancellation status
- Updated ResponsePaneHeader with cancellation message support

### Detailed Implementation Changes

#### packages/apidash_core/lib/services/http_service.dart
- Added specific exception handling for request cancellation:
```dart
if (e.message.contains('Connection closed') || e.message.contains('abort')) {
  return (null, null, 'Request Cancelled');
} else {
  return (null, null, e.toString());
}
```

#### lib/providers/collection_provider.dart
- Introduced client management with a map to store HTTP clients:
```dart
final Map<String, http.Client> _clients = {};
```
- Enhanced sendRequest method:
```dart
final client = http.Client();
_clients[id] = client;
```
- Added cancelRequest method:
```dart
void cancelRequest(String id) {
  _clients[id]?.close();
  _clients.remove(id);
  // Update the request state to reflect cancellation
}
```

#### lib/screens/home_page/editor_pane/details_card/response_pane.dart
- Added cancellation UI handling:
```dart
if (message == "Request Cancelled") {
  return ErrorMessage(
    message: '$message',
    showIssueButton: false,
  );
}
```

#### lib/screens/home_page/editor_pane/url_card.dart
- Updated SendRequestButton with cancellation support:
```dart
return SendButton(
  isWorking: isWorking ?? false,
  onTap: () {
    ref.read(collectionStateNotifierProvider.notifier).sendRequest(selectedId!);
  },
  onCancel: () {
    ref.read(collectionStateNotifierProvider.notifier).cancelRequest(selectedId!);
  },
);
```

#### lib/widgets/button_send.dart
- Updated SendRequestButton with cancellation support.

#### lib/widgets/response_widgets.dart
- Added cancellation message handling:
```dart
if (message == "Request Cancelled") {
  return ErrorMessage(
    message: '$message',
    showIssueButton: false,
  );
}
```

ashitaprasad commented on lib/providers/collection_providers.dart:
```diff
@@ -1,6 +1,7 @@
 import 'package:apidash_core/apidash_core.dart';
 import 'package:flutter_riverpod/flutter_riverpod.dart';
 import 'package:apidash/consts.dart';
+import 'package:http/http.dart' as http;
```
direct import of http package is not allowed .. all http activities must be done in apidash_core.

ashitaprasad commented on lib/providers/collection_providers.dart:
```diff
@@ -46,6 +47,7 @@ class CollectionStateNotifier
   final Ref ref;
   final HiveHandler hiveHandler;
   final baseResponseModel = const HttpResponseModel();
+  final Map<String, http.Client> _clients = {};
```
Must not be a field inside a StateNotifierProvider whose goal is to store the Map<String, RequestModel>?
leads to mingling of data.

ashitaprasad commented on lib/providers/collection_providers.dart:
```diff
@@ -259,9 +261,14 @@ class CollectionStateNotifier
 
+    // Create an HTTP client and store it to allow cancellation
+    final client = http.Client();
+    _clients[id] = client;
```
all http activities should be inside apidash_core.

ashitaprasad commented on lib/providers/collection_providers.dart:
```diff
@@ -308,6 +319,24 @@ class CollectionStateNotifier
+      ref.read(hasUnsavedChangesProvider.notifier).state = true;
+    }
+  }
+
```
all http activities inside apidash_core.
Also, isn't _clients map and isWorking field kind of redundant now as all working request ids are present in _clients.

ashitaprasad commented on lib/screens/home_page/editor_pane/details_card/response_pane.dart:
```diff
@@ -27,6 +27,14 @@ class ResponsePane extends ConsumerWidget {
+        showIssueButton: false,
+      );
+    }
+
```
responseStatus is null so this new code addition is unreachable.
Also, not required as cancelling a request is not an error ..

ashitaprasad commented on lib/widgets/response_widgets.dart:
```diff
@@ -140,7 +140,9 @@ class ResponsePaneHeader extends StatelessWidget {
-                "$responseStatus: ${message ?? '-'}",
+                message == 'Request Cancelled'
+                    ? 'Request Cancelled'
+                    : "$responseStatus: ${message ?? '-'}",
```
not required

ashitaprasad commented on packages/apidash_core/lib/services/http_service.dart:
```diff
@@ -68,29 +74,42 @@ Future<(HttpResponse?, Duration?, String?)> request(
       return (response, stopwatch.elapsed, null);
+    } on http.ClientException catch (e) {
+      if (e.message.contains('Connection closed') ||
+          e.message.contains('abort')) {
```
What is the source of this info that these strings - 'Connection closed' & 'abort' will occur?

sasanktumpati commented:
## Changes

1. **Created a new `HttpClientManager` singleton in `apidash_core` to manage HTTP clients:**

   ```dart
   // filepath: packages/apidash_core/lib/services/client_manager.dart
   class HttpClientManager {
     static final HttpClientManager _instance = HttpClientManager._internal();
     final Map<String, http.Client> _clients = {};

     // Create and track client for each request
     http.Client createClient(String requestId) { }

     // Cancel and cleanup client
     void cancelRequest(String requestId) { }
   }
   ```

2. **Updated HTTP service to use `HttpClientManager`:**
3. **Updated Tests for the Send Button**

sasanktumpati commented:
Hi @ashitaprasad,

I’ve made the updates based on your suggestions, including implementing the `HttpClientManager` singleton to handle HTTP clients and updating the HTTP service to use it. The tests for the Send Button have also been adjusted accordingly.

Regarding the strings 'Connection closed' and 'abort' in the exception handling: I went through the documentation but didn’t find any specific mention of these strings. So, I added them based on my own testing. 

Let me know if you have any further feedback!

ashitaprasad commented on lib/providers/collection_providers.dart:
```diff
@@ -237,21 +237,15 @@ class CollectionStateNotifier
-        (value) => value.defaultUriScheme,
-      ),
+      settingsProvider.select((value) => value.defaultUriScheme),
     );
```
unnecessary code change

ashitaprasad commented on lib/providers/collection_providers.dart:
```diff
@@ -237,21 +237,15 @@ class CollectionStateNotifier
-    if (requestModel.httpRequestModel == null) {
-      return;
-    }
+    if (requestModel.httpRequestModel == null) return;
```
unnecessary code change

ashitaprasad commented on lib/providers/collection_providers.dart:
```diff
@@ -300,14 +304,29 @@ class CollectionStateNotifier
     ref.read(hasUnsavedChangesProvider.notifier).state = true;
   }
 
+  final httpClientManager = HttpClientManager();
```
Why httpClientManager is being initialized inside the statenotifiedprovider

ashitaprasad replied:
That too in between methods

ashitaprasad commented on lib/providers/collection_providers.dart:
```diff
@@ -300,14 +304,29 @@ class CollectionStateNotifier
+      responseStatus: null,
+      httpResponseModel: null,
+    );
+    state = map;
```
If cancelled request is being marked here .. why the changes of L264-276 required

ashitaprasad commented on lib/screens/home_page/editor_pane/details_card/response_pane.dart:
```diff
@@ -27,6 +27,7 @@ class ResponsePane extends ConsumerWidget {
     if (responseStatus == null) {
       return const NotSentWidget();
     }
+
```
unnecessary code change

ashitaprasad commented on packages/apidash_core/lib/services/http_service.dart:
```diff
@@ -68,29 +82,45 @@ Future<(HttpResponse?, Duration?, String?)> request(
+        return (null, null, 'Request Cancelled');
+      } else {
+        return (null, null, e.toString());
+      }
```
This entire logic still doesn't make sense ... if a user explicitly closes connection .. Whatever be the exception here .. the message is set inside the `cancelRequest(String id)` inside the statenotifierprovider

ashitaprasad commented:
@sasanktumpati stop marking comments as resolved. The reviewer will review and mark it as resolved. You need to just add a comment that this comment has been addressed.

sasanktumpati commented:
> @sasanktumpati stop marking comments as resolved. The reviewer will review and mark it as resolved. You need to just add a comment that this comment has been addressed.

sorry for that

ashitaprasad commented:
> Regarding the strings 'Connection closed' and 'abort' in the exception handling: I went through the documentation but didn’t find any specific mention of these strings. So, I added them based on my own testing.

This should have been added as a comment to review comment where I had asked the question.

sasanktumpati commented:
> > Regarding the strings 'Connection closed' and 'abort' in the exception handling: I went through the documentation but didn’t find any specific mention of these strings. So, I added them based on my own testing.
> 
> This should have been added as a comment to review comment where I had asked the question.

![image](https://github.com/user-attachments/assets/45b769bc-28ab-4fc1-8488-c855cc803c57)
when we close a client, this is the error message

ashitaprasad commented:
> > > Regarding the strings 'Connection closed' and 'abort' in the exception handling: I went through the documentation but didn’t find any specific mention of these strings. So, I added them based on my own testing.
> > 
> > 
> > This should have been added as a comment to review comment where I had asked the question.
> 
> ![image](https://private-user-images.githubusercontent.com/76063158/393501230-45b769bc-28ab-4fc1-8488-c855cc803c57.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM1NjYxMjIsIm5iZiI6MTczMzU2NTgyMiwicGF0aCI6Ii83NjA2MzE1OC8zOTM1MDEyMzAtNDViNzY5YmMtMjhhYi00ZmMxLTg0ODgtYzg1NWNjODAzYzU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA3VDEwMDM0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlMmVkMWUyMGVjYTEwNDQyMzdkZTk5ODMzZTRhNTY1NmM2ZjExZmNkMjYyY2Q5ZGU3YzI4MmVmODI0Mjk4MTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ci2JxobnG4ItUo44YFApR6gd-8XWVJvw_LfAAWF0hh8) when we close a client, this is the error message

Which means that the state changes being made in this function is redundant and not having any impact
```
void cancelRequest(String id) {
    httpClientManager.cancelRequest(id);
    var currentModel = state![id]!;
    var map = {...state!};
    map[id] = currentModel.copyWith(
      isWorking: false,
      message: 'Request Cancelled',
      responseStatus: null,
      httpResponseModel: null,
    );
    state = map;
```

sasanktumpati commented:
> > > > Regarding the strings 'Connection closed' and 'abort' in the exception handling: I went through the documentation but didn’t find any specific mention of these strings. So, I added them based on my own testing.
> > > 
> > > 
> > > This should have been added as a comment to review comment where I had asked the question.
> > 
> > 
> > ![image](https://private-user-images.githubusercontent.com/76063158/393501230-45b769bc-28ab-4fc1-8488-c855cc803c57.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM1NjYxMjIsIm5iZiI6MTczMzU2NTgyMiwicGF0aCI6Ii83NjA2MzE1OC8zOTM1MDEyMzAtNDViNzY5YmMtMjhhYi00ZmMxLTg0ODgtYzg1NWNjODAzYzU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA3VDEwMDM0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlMmVkMWUyMGVjYTEwNDQyMzdkZTk5ODMzZTRhNTY1NmM2ZjExZmNkMjYyY2Q5ZGU3YzI4MmVmODI0Mjk4MTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ci2JxobnG4ItUo44YFApR6gd-8XWVJvw_LfAAWF0hh8) when we close a client, this is the error message
> 
> Which means that the state changes being made in this function is redundant and not having any impact
> 
> ```
> void cancelRequest(String id) {
>     httpClientManager.cancelRequest(id);
>     var currentModel = state![id]!;
>     var map = {...state!};
>     map[id] = currentModel.copyWith(
>       isWorking: false,
>       message: 'Request Cancelled',
>       responseStatus: null,
>       httpResponseModel: null,
>     );
>     state = map;
> ```

yeah, fixed that.

ashitaprasad commented:
Any updates @sasanktumpati

sasanktumpati commented:
> Any updates @sasanktumpati 

I have fixed then in the previous commit, now have to figure out something for the strings comparison

ashitaprasad commented:
> > Any updates @sasanktumpati
> 
> I have fixed then in the previous commit, now have to figure out something for the strings comparison

Cool .. 👍

sasanktumpati commented:
@ashitaprasad you can review it now

ashitaprasad commented on lib/providers/collection_providers.dart:
```diff
@@ -262,14 +264,25 @@ class CollectionStateNotifier
+          message: kMsgRequestCancelled,
+          isWorking: false,
+          httpResponseModel: null,
+        );
```
After changes, responseRec.$3 is returned as  kMsgRequestCancelled nowhere in the http_service, so how do you know that this condition is satisfied and the request is cancelled by the user.

sasanktumpati commented:
Hi @ashitaprasad,

Apologies again for the earlier confusion. Here's a summary of the latest commit:

- **Cancellation Tracking**:
  - Introduced a `Queue<String>` to track cancelled request IDs.
  - Implemented a 'First In, First Out' approach with a maximum of 100 entries to avoid memory overhead.
  - Automatically removes older entries when the queue reaches its limit.

- **HTTP Service**:
  - Integrated the queue to directly check if a request was cancelled within the `request` function.
  - Returns `"Request cancelled"` when a cancelled request is detected.

ashitaprasad commented:
Hi @sasanktumpati Thanks for picking up this long pending issue and taking it to the finish line.
Congratulations on you first awesome contribution to API Dash 🚀
